### ISSUE TITLE
ER: CSV/TSV/JSON output using console.sh

### ISSUE THREAD

**Issue by pkoppstein** [orientechnologies/orientdb#5435]
One of the reasons for the universal success of tools like awk,
SQLite, and R is that they support easy data-interoperability,
and I think that with little effort on the part of OrientDB developers,
OrientDB would gain a great deal of traction with a few enhancements.

In this ER, I'll focus just on the "output" side of things, i.e.
enhancements to support the use of console.sh (or console.bat) to produce output
in CSV, TSV, and JSON formats. 

The first and most important step would be to support the functionality offered by sqlite3's commands:

```
.headers
.mode     # In particular, the csv, tabs, and list subcommands
.once     # Output to FILENAME for the next SQL command only
.output
.print
.separator
```

The second step would be to support JSON more comprehensively in the "console.sh" environment. To explain what I mean, consider the following typescript:

```
orientdb {db=demo}> select expand(out()) from #16:0

----+-----+------+----+--------------
#   |@RID |@CLASS|name|in_IsNeighbour
----+-----+------+----+--------------
0   |#16:1|Person|P2  |[#17:1]       
1   |#16:2|Person|P3  |[#17:2]       
----+-----+------+----+--------------

orientdb {db=demo}> select @this.toJSON() from (select expand(out()) from #16:0)

----+------+-----------------------------------------------------------------------------------------------------------------------------------
#   |@CLASS|this                                                                                                                               
----+------+-----------------------------------------------------------------------------------------------------------------------------------
0   |null  |{"@type":"d","@rid":"#16:1","@version":2,"@class":"Person","name":"P2","in_IsNeighbour":["#17:1"],"@fieldTypes":"in_IsNeighbour=g"}
1   |null  |{"@type":"d","@rid":"#16:2","@version":2,"@class":"Person","name":"P3","in_IsNeighbour":["#17:2"],"@fieldTypes":"in_IsNeighbour=g"}
----+------+-----------------------------------------------------------------------------------------------------------------------------------

orientdb {db=demo}> select @this["version"] from (select @this.toJSON() from (select expand(out()) from #16:0))

----+------
#   |@CLASS
----+------
0   |null  
1   |null  
----+------
```

(Using "@this.version" produces the same results.)

In short, it appears that .toJSON() returns a string, from which it is
not possible to extract information using the usual notational
devices.

Finally, I'd like to request that support for CSV and TSV output be provided with
the same level of attention to detail as exemplified by the @csv and @tsv filters in jq Version 1.5
These filters do a fantastic job of dealing with edge-cases (e.g. NUL).
See https://stedolan.github.io/jq/manual/

**Comment by pkoppstein** [comment]
The cry for help (with notes of exasperation) at https://stackoverflow.com/questions/34279189/insert-50k-json-objects-into-database could be taken as an indicator of the urgency of this ER.  Or maybe I should say that by "easy data-interoperability" I did mean to imply that performant data-interoperability should also be easy to attain :-)

**Comment by smolinari** [comment]
The SO thread, as I read it, is about performance and not necessarily about the missing ability of importing data formats. 

Importing JSON, XML or CSV data is possible with ETL.

http://orientdb.com/docs/2.1/Extractor.html

On the exporting side, ODB only supports JSON output. 

Scott

**Comment by pkoppstein** [comment]
See also https://stackoverflow.com/questions/34446497/orientjs-how-to-get-standard-json-serialized-from-query

**Comment by robfrank** [comment]
After some analysis, to allow a different output format we need to deeply redesign the console code. 

About
```
select @this["version"] from (select @this.toJSON() from (select expand(out()) from #16:0))
```
this means embed a json parser inside the console itself to post-process the result set: non feasabile
